### PR TITLE
Examples code updated with changes as requested (default values removed, use of variables restricted to only least, local values used wherever applicable, provider with lowest and one latest version defined, test/plan/apply verified)

### PR DESCRIPTION
**Pull request by marifse**
### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Comment by marifse** [comment]
@ocofaigh can you please check and approve if okay.

**Review by ocofaigh — CHANGES_REQUESTED** [review]
There is only a basic example here which tests the lowest provider version. Can you create an "advanced" example which will test latest provider version. It should support the following:
- resource group creation
- key protect instance with
  - `service_endpoints` set to private
  - pass some values for `key_create_import_access_settings`
- create new key ring (using https://github.com/terraform-ibm-modules/terraform-ibm-kms-key-ring)
- create new key in new key ring (https://github.com/terraform-ibm-modules/terraform-ibm-kms-key)

You should add a new test for the new example in pr_test.go and you should update the cra-config.yaml to now scan the new example

**Comment by marifse** [comment]
> st latest provider version. It should support the following:

sure. let me

**Comment by marifse** [comment]
> There is only a basic example here which tests the lowest provider version. Can you create an "advanced" example which will test latest provider version. It should support the following:
> 
> * resource group creation
> * key protect instance with
>   
>   * `service_endpoints` set to private
>   * pass some values for `key_create_import_access_settings`
> * create new key ring (using https://github.com/terraform-ibm-modules/terraform-ibm-kms-key-ring)
> * create new key in new key ring (https://github.com/terraform-ibm-modules/terraform-ibm-kms-key)
> 
> You should add a new test for the new example in pr_test.go and you should update the cra-config.yaml to now scan the new example

@ocofaigh can you please check now, i have added the advance example, tested it, though i have tested it with public access as for private i have to be IBM cloud private network. So i added a comment in main.tf, you can review and let me know your thoughts. Also i have added the test in pre-test.go but i do not try running that.

**Comment by marifse** [comment]
@ocofaigh i have a query, if i am removing some variable default value, then i am supposed to add the variable in catalogValidationValues.json.template.json, with what value "hard coded" or like this $variablevalue, like in another PGSQL example, i removed the default value for admin_pass variable, and i have to include it in catalogValidationValues.json.template.json, it below is okay? and if okay where do i need to pass the value so that the test can run successfull.

{
  "ibmcloud_api_key": $VALIDATION_APIKEY,
  "region": "us-south",
  "resource_tags": $TAGS,
  "prefix": $PREFIX,
  "admin_pass": $ADMIN_PASSWORD
}

Also where are the values for these $ are getting fetched when running the tests, or i need to pass them in test option.

File: examples/advance/main.tf
Comment (ocofaigh) [inline examples/advance/main.tf:24]: set this to `service_endpoints = "private"` - we can write a test that actually tests this test in IBM Schematics which will have access to the private network (see subsequent comment)

File: examples/advance/version.tf
Comment (ocofaigh) [inline examples/advance/version.tf:4]: This comment is not valid. Please update to:

```
# Ensure that there is always 1 example locked into the lowest provider version of the range defined in the main
# module's version.tf (basic example), and 1 example that will always use the latest provider version (advanced example).
```

File: tests/pr_test.go
Comment (ocofaigh) [inline tests/pr_test.go:58]: So in order to test the advanced example, lets copy the approach [here](https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/blob/0295160a963d3beb60394bfd60da6a42513f83f1/tests/pr_test.go#L87-L127) which will deploy the example in schematics which means the private endpoint can be used.

File: examples/advance/README.md
Comment (ocofaigh) [inline examples/advance/README.md:3]: An advanced example that shows how to provision a private only Key Protect instance with a new Key Ring and Key.

File: examples/advance/README.md
Comment (ocofaigh) [inline examples/advance/README.md:7]: Private -> private

File: examples/advance/main.tf
Comment (ocofaigh) [inline examples/advance/main.tf:14]: remove "and Keyring"

File: examples/advance/main.tf
Comment (ocofaigh) [inline examples/advance/main.tf:35]: use the prefix for naming. EG: `"${var.prefix}-my-key-ring"`

File: examples/advance/main.tf
Comment (ocofaigh) [inline examples/advance/main.tf:36]: Don't use `depends_on` - there is already an implicit dependency with the instance GUID

**Review by ocofaigh — CHANGES_REQUESTED** [review]
I left some comments. Could you also please add the new advanced exampled to the ibm_catalog.json in the flavors array here -> https://github.com/marifse/terraform-ibm-key-protect/blob/52f7bb3daafe237c9c5ceb81961680067d366b15/ibm_catalog.json#L54

In terms of the variables in the `catalogValidationValues.json.template.json` file, they are not used by the PR pipeline, they are used by an internal pipeline AFTER code is merged. I can assist you with any new required values as part of the PR review process. Unfortunately the full list of them are only documented internally since its an internal only pipeline.

File: examples/advance/main.tf
Comment (marifse) [reply examples/advance/main.tf]: @ocofaigh requested changes have been addresses, except the pre-test.go, i am not sure either that is correct or not. can you please look into that and let me know further. thanks

File: tests/pr_test.go
Comment (ocofaigh) [inline tests/pr_test.go:85]: There is currently a provider bug which this test will hit when using private only instance. To workaround it, please add this:

```
// Workaround for https://github.com/IBM-Cloud/terraform-provider-ibm/issues/5154
options.AddWorkspaceEnvVar("IBMCLOUD_KP_API_ENDPOINT", "https://private."+region+".kms.cloud.ibm.com", false, false)
```

**Review by ocofaigh — CHANGES_REQUESTED** [review]
see comments

File: examples/advance/README.md
Comment (marifse) [reply examples/advance/README.md]: done

**Comment by marifse** [comment]
> I left some comments. Could you also please add the new advanced exampled to the ibm_catalog.json in the flavors array here -> https://github.com/marifse/terraform-ibm-key-protect/blob/52f7bb3daafe237c9c5ceb81961680067d366b15/ibm_catalog.json#L54
> 
> In terms of the variables in the `catalogValidationValues.json.template.json` file, they are not used by the PR pipeline, they are used by an internal pipeline AFTER code is merged. I can assist you with any new required values as part of the PR review process. Unfortunately the full list of them are only documented internally since its an internal only pipeline.

@ocofaigh please check now.

File: tests/pr_test.go
Comment (ocofaigh) [inline tests/pr_test.go:81]: There is no need for this logic here as we are not using HPCS in this example

**Review by ocofaigh — CHANGES_REQUESTED** [review]
Hmm, the golang linter is failing in the pipeline. Are you able to reproduce it locally by running `pre-commit run --all-files ` ?

Here is what pipeline is failing with: `level=error msg="Running error: can't run linter goanalysis_metalinter\nbuildir: failed to load package : could not load export data: no export data for \"github.com/IBM/schematics-go-sdk/schematicsv1\""`

File: tests/pr_test.go
Comment (marifse) [reply tests/pr_test.go]: @ocofaigh removed this logic

Comment (marifse) [reply tests/pr_test.go]: @ocofaigh do have any clue on golang_lint, how can i fix it. It is also failing on my side even after updating the golang_lint

File: tests/pr_test.go
Comment (ocofaigh) [inline tests/pr_test.go:9]: @marifse maybe try moving this as the last import? Just a guess, but that works in [this test](https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/blob/bcdb164522ff654f674fa7bef717a97f5b1aa5f9/tests/pr_test.go#L4-L13)

**Comment by ocofaigh** [comment]
@marifse It looks like you removed the test in your latest commit - we need a test for the advanced example

**Comment by marifse** [comment]
@ocofaigh test added for advanced example again, and also the go_lint error fixed. Please review and re-run the pipeline.

File: tests/pr_test.go
Comment (marifse) [reply tests/pr_test.go:75]: @ocofaigh after adding this line, i started to get the below error:

pr_test.go:1: : # github.com/terraform-ibm-modules/terraform-ibm-key-protect [github.com/terraform-ibm-modules/terraform-ibm-key-protect.test]
./pr_test.go:77:76: undefined: region (typecheck)

Before this the CI ran successful

Comment (marifse) [reply tests/pr_test.go:75]: @ocofaigh can you please try running the pipeline now.

Comment (ocofaigh) [reply tests/pr_test.go:75]: @marifse thats because you don't have any region variable defined. Try this:
```
// Workaround for https://github.com/IBM-Cloud/terraform-provider-ibm/issues/5154 options.AddWorkspaceEnvVar("IBMCLOUD_KP_API_ENDPOINT", "https://private."+options.Region+".kms.cloud.ibm.com", false, false)
```

Comment (marifse) [reply tests/pr_test.go:75]: @ocofaigh done. I also tried that but that time it did not work at first. Anyways now it is okay.

File: tests/pr_test.go
Comment (ocofaigh) [inline tests/pr_test.go:66]: there is no modules folder in this repo so I think thats why test failed in pipeline. You will want to add `advancedExampleTerraformDir` here

Comment (marifse) [reply tests/pr_test.go]: @removed this folder. please re-run

File: tests/pr_test.go
Comment (ocofaigh) [inline tests/pr_test.go:65]: you need to also add the example directory (aka `advancedExampleTerraformDir`)

Comment (marifse) [reply tests/pr_test.go]: @ocofaigh done

File: tests/pr_test.go
Comment (ocofaigh) [inline tests/pr_test.go:65]: please add back the root level - the example calls the root level module so both are needed in the tar file that gets sent to schematics
```
TarIncludePatterns: []string{
			"*.tf",
			advancedExampleTerraformDir + "/*.tf",
		},
```

Comment (marifse) [reply tests/pr_test.go:66]: @ocofaigh done.

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**
